### PR TITLE
Define initial CI

### PR DESCRIPTION
**Pull request by Sinclert**
This PR defines the proposed pair of initial _GitHub Actions_ CI jobs, based on the `Makefile` rules that [PR #1](https://github.com/HEPData/hepdata-pyhf-extractor/pull/1) will introduce.

I created this PR separated from [PR #1](https://github.com/HEPData/hepdata-pyhf-extractor/pull/1), in case further discussion occurs.

#### Considerations
- Running the format checker with Python3.6 (I think it does not matter as far as Python3 is used).
- Running the test job on the Python3 versions currently available (3.6, 3.7 and 3.8).

#### Future work
- If HEPData team likes GitHub Actions, future iterations could include other workflows to build and publish particular versions to PyPi (through the use of [secrets](https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets)).

### CODE REVIEW

**Review by kratsg — COMMENTED** [review]
Looks ok to me, but I will defer to @matthewfeickert 's judgement here. He knows specifically what is needed to get things working with python in GHA perfectly.

File: .github/workflows/ci.yml
Comment (matthewfeickert) [inline .github/workflows/ci.yml:6]: You probably want CI to run on every push, no? Not just on PRs?
```suggestion
```

File: .github/workflows/ci.yml
Comment (matthewfeickert) [inline .github/workflows/ci.yml:14]: Add nightly CRON with schedule and on demand runs with workflow dispatch

```suggestion
  # Run daily at 0:01 UTC
  schedule:
  - cron:  '1 0 * * *'
  workflow_dispatch:
```

File: .github/workflows/ci.yml
Comment (matthewfeickert) [inline .github/workflows/ci.yml]: I've come to favor factoring out things like linting into a separate workflow so that things can be run individually if needed using workflow dispatch. c.f. [`pyhf`'s `lint` workflow](https://github.com/scikit-hep/pyhf/blob/master/.github/workflows/lint.yml).

File: .github/workflows/ci.yml
Comment (matthewfeickert) [inline .github/workflows/ci.yml]:
```suggestion
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        os: [ubuntu-latest, macos-latest]
        python-version: [3.6, 3.7, 3.8]
```

Maybe as the point of this to be deployed on HEPData it really only matters if this runs on Linux, but I like to test MacOS as well usually for the actual pytest suite.

File: .github/workflows/ci.yml
Comment (matthewfeickert) [inline .github/workflows/ci.yml]:
```suggestion
        python -m pip install --upgrade pip setuptools wheel
        python -m pip install --ignore-installed --quiet --no-cache-dir .[dev]
        python -m pip list
```

I usually like to upgrade `pip`, `setuptools`, and `wheel` so that the CI is testing what a brand new user would probably get and then in addition to a few small things `pip list` at the end so that we can easily see in the logs between runs if there are dependency version changes.

**Review by matthewfeickert — COMMENTED** [review]
Thanks for getting this going @Sinclert. Just some comments from me, but again this is fine as is.

File: .github/workflows/ci.yml
Comment (Sinclert) [reply .github/workflows/ci.yml:6]: m... m... yes, on every push **to master**, not on every push to _\<random-development-branch\>_.

If users want to check format, or test compliance before proposing a PR, they can just use the `Makefile` rules. These CI jobs are designed to be triggered **just before** a branch gets merged to `master` (or when a punctual fix gets directly merged to `master`).

File: .github/workflows/ci.yml
Comment (Sinclert) [reply .github/workflows/ci.yml:14]: I am ok with `workflow_dispatch`, as it is something convenient.

Now, when does a cron-job CI run going to detect issues (in format, or test compliance), that neither the PR creation triggers, nor the `master` push trigger detected?

File: .github/workflows/ci.yml
Comment (Sinclert) [reply .github/workflows/ci.yml]: Same comment as previous suggestion:

When does a format-check CI run going to detect issues, that neither the PR creation triggers, nor the master push trigger detected?

Format could be mess up on a particular development branch, and that is OK. It only becomes relevant when it is proposed to be merged to `master`, and the [pre-commit hook](https://github.com/HEPData/hepdata-pyhf-extractor/blob/2fe51eebd92c3103e32b52013a7f53be843e729d/.pre-commit-config.yaml) was not installed (for some reason).

File: .github/workflows/ci.yml
Comment (Sinclert) [reply .github/workflows/ci.yml]: I am fine with adding macOS.

File: .github/workflows/ci.yml
Comment (Sinclert) [reply .github/workflows/ci.yml]: I am fine with this change.

Although, please consider that installing `setup.py` particular set of dependencies (in this case`.[dev]`) without quotes (`".[dev]"`), does not work on certain shells (i.e. [zsh](https://ohmyz.sh/)).

File: .github/workflows/ci.yml
Comment (matthewfeickert) [reply .github/workflows/ci.yml:6]: Okay, I guess the `pyhf` team probably spams the CI systems then a we run on every dev branch. They're so lightweight that we've never seen a reason to not run them. But this isn't a big deal here, so let's drop this and move on.

File: .github/workflows/ci.yml
Comment (matthewfeickert) [reply .github/workflows/ci.yml:14]: You'd want to know ahead of time if changes in dependency releases are going to affect you.

**Review by GraemeWatt — APPROVED** [review]
This looks good to me.  I'll let you CI experts argue about the details.  I suppose the reason for a nightly cron job would be to check for a new release of a dependent package which breaks the build or tests.  But this might be overkill for a simple repository with minimal dependencies.

File: .github/workflows/ci.yml
Comment (matthewfeickert) [reply .github/workflows/ci.yml]: sure, but the CI machines don't run `zsh`. This is good to know though, as I didn't know `zsh` was that picky. :(

**Comment by GraemeWatt** [comment]
> #### Future work
> * If HEPData team likes GitHub Actions, future iterations could include other workflows to build and publish particular versions to PyPi (through the use of [secrets](https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets)).

I generated API tokens from the `hepdata` PyPI account and added `PYPI_PASSWORD` and `TEST_PYPI_PASSWORD` as GitHub Secrets, following "[Publishing package distribution releases using GitHub Actions CI/CD workflows](https://packaging.python.org/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/)".  So in another PR you could add a GitHub Actions workflow to automatically deploy tagged releases on PyPI.

File: .github/workflows/ci.yml
Comment (matthewfeickert) [reply .github/workflows/ci.yml]: > Format could be mess up on a particular development branch, and that is OK. It only becomes relevant when it is proposed to be merged to master, and the pre-commit hook was not installed (for some reason).

I would still advocate for factorization of workflows for different purposes. We used to have everything be in one big GHA workflow file for `pyhf`, but as time has moved on we've factored more and more of it out into separate workflows &mdash; this has worked really well and dropped the entire CI time. While this does introduce additional boilerplate at the top of each workflow file, it is nice to be able to just rerun more specific workflows if there are errors on GitHub's side that cause the workflow to fail. Additionally, if you require all the checks to pass in GitHub branch protections options then you can run all workflows in parallel. Here, this saves almost no times as everything is so small and so fast, but at the same time why not?

**Comment by matthewfeickert** [comment]
> I generated API tokens from the `hepdata` PyPI account and added `PYPI_PASSWORD` and `TEST_PYPI_PASSWORD` as GitHub Secrets, following "[Publishing package distribution releases using GitHub Actions CI/CD workflows](https://packaging.python.org/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/)". So in another PR you could add a GitHub Actions workflow to automatically deploy tagged releases on PyPI.

Nice! Thanks @GraemeWatt. Automatic publishing is a bit weird, and we do it in a pretty niche way for `pyhf`, but @lukasheinrich and I setup what seems to be a pretty reasonable workflow for `pylhe` that publishes though a mixture of GitHub actions triggered by GitHub releases. c.f. https://github.com/scikit-hep/pylhe/pull/44 (though there were some fixes following that).

File: .github/workflows/ci.yml
Comment (Sinclert) [reply .github/workflows/ci.yml]: The main reason to avoid running different jobs in parallel (`check` and `test`), is that one can potentially get more computational expensive than the other one (in this case, `test`). So, you will only want to trigger that job when you are absolutely sure the previous _light-weight_ jobs are successful.

If you move the `check` / `lint` job into another workflow, you are unable to reference it in the `needs: [<job_ID>]` section, and therefore, you will end up spawning `test` jobs without even checking if the style is correct.

This _not-doing-lightweights-checks-first_ methodology could even get more inefficient if a type checker like [Mypy](http://mypy-lang.org/) is introduced. Mypy checks would run faster than a proper test suite run, so it would become another _lightweight_ check that, hopefully, could be referenced in the `needs` section of the `test` job.

File: .github/workflows/ci.yml
Comment (Sinclert) [reply .github/workflows/ci.yml:14]: That is a good point!

I think that is useful when there is the possibility that **installation** requirements could break. Right now, [PR #1](https://github.com/HEPData/hepdata-pyhf-extractor/pull/1) does not introduce any _installation_ requirements, only _development_ ones.

We could then setup this cronjob rule for the future, although I am not sure what installation requirements may come down the way. Perhaps `pyyaml`? 🤷🏻‍♂️

File: .github/workflows/ci.yml
Comment (matthewfeickert) [reply .github/workflows/ci.yml]: Okay. It just sounds like you're wanting to optimize the CI differently than I would, which is fine, so I'd just move forward with what you have. 

For `pyhf` we optimize for running all workflows in the shortest amount of wall time for PR given that you can require all checks pass through branch protection options. You don't want runs blocked by linting if you care more about making sure if the test suite is going to pass or not.

**Review by matthewfeickert — APPROVED** [review]

**Pull request merged**
